rewrite-java/src/main/java/org/openrewrite/java/internal/beta/CallMatcher.java

```diff
@@ -0,0 +1,133 @@
+import org.openrewrite.java.tree.Expression;
+import org.openrewrite.java.tree.Flag;
+import org.openrewrite.java.tree.J;
+import org.openrewrite.java.tree.JavaType;
+
+import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+@FunctionalInterface
+public interface CallMatcher {
+    boolean matches(Expression expression);
+
+    default AdvancedCallMatcher advanced() {
+        return new AdvancedCallMatcher(this);
+    }
+
+    static CallMatcher fromMethodMatcher(MethodMatcher methodMatcher) {
+        return methodMatcher::matches;
+    }
+
+    static CallMatcher fromCallMatchers(Collection<CallMatcher> matchers) {
+        if (matchers.size() > 750) {
+            return expression -> matchers.parallelStream().anyMatch(matcher -> matcher.matches(expression));
+        } else {
+            return expression -> matchers.stream().anyMatch(matcher -> matcher.matches(expression));
+        }
+    }
+
+    static CallMatcher fromCallMatchers(MethodMatcher... methodMatchers) {
+        return fromCallMatchers(Stream.of(methodMatchers).map(CallMatcher::fromMethodMatcher).collect(Collectors.toList()));
+    }
+
+    static CallMatcher fromMethodMatchers(Collection<MethodMatcher> matchers) {
+        return fromCallMatchers(matchers.stream().map(CallMatcher::fromMethodMatcher).collect(Collectors.toList()));
+    }
+
+    @AllArgsConstructor(access = AccessLevel.PRIVATE)
+    class AdvancedCallMatcher {
+        CallMatcher matcher;
+
+        public boolean isSelect(Expression expression, Cursor cursor) {
+            assert expression == cursor.getValue() : "expression != cursor.getValue()";
+            J.MethodInvocation maybeMethodInvocation =
+                    cursor.getParentOrThrow().firstEnclosing(J.MethodInvocation.class);
+            return maybeMethodInvocation != null &&
+                    maybeMethodInvocation.getSelect() == expression &&
+                    matcher.matches(maybeMethodInvocation); // Do the matcher.matches(...) last as this can be expensive
+        }
+
+        public boolean isAnyArgument(Expression expression, Cursor cursor) {
+            assert expression == cursor.getValue() : "expression != cursor.getValue()";
+            J stop = cursor.dropParentUntil(v -> v instanceof J.MethodInvocation || v instanceof J.NewClass || v instanceof J.Block).getValue();
+            if (stop instanceof J.Block) {
+                return false;
+            }
+            Expression call = (Expression) stop;
+            return getCallArguments(call).contains(expression) && matcher.matches(call); // Do the matcher.matches(...) last as this can be expensive
+        }
+
+        public boolean isFirstParameter(Expression expression, Cursor cursor) {
+            return isParameter(expression, cursor, 0);
+        }
+
+        public boolean isParameter(Expression expression, Cursor cursor, int parameterIndex) {
```

Comment:
should this be `isMethodOrNewClassParameter`?

Reply:
It's a method on `AdvancedCallMatcher`. I think that the name is implied "a parameter on a call"